substrate/network/src/consensus.rs

```diff
@@ -0,0 +1,190 @@
+
+use std::collections::HashMap;
+use multiqueue;
+use futures::sync::oneshot;
+use io::SyncIo;
+use protocol::Protocol;
+use network::PeerId;
+use primitives::Hash;
+use message::{self, Message};
+use runtime_support::Hashable;
+
+struct CandidateRequest {
+	id: message::RequestId,
+	completion: oneshot::Sender<Vec<u8>>,
+}
+
+struct PeerConsensus {
+	candidate_fetch: Option<CandidateRequest>,
+	candidate_available: Option<Hash>,
+}
+
+/// Consensus network protocol handler. Manages statements and candidate requests.
+pub struct Consensus {
+	peers: HashMap<PeerId, PeerConsensus>,
+	our_candidate: Option<(Hash, Vec<u8>)>,
+	statement_sink: multiqueue::BroadcastFutSender<message::Statement>,
+	statement_stream: multiqueue::BroadcastFutReceiver<message::Statement>,
+	bft_message_sink: multiqueue::BroadcastFutSender<message::BftMessage>,
+	bft_message_stream: multiqueue::BroadcastFutReceiver<message::BftMessage>,
+}
+
+impl Consensus {
+	/// Create a new instance.
+	pub fn new() -> Consensus {
+		let (statement_sink, statement_stream) = multiqueue::broadcast_fut_queue(64);
+		let (bft_sink, bft_stream) = multiqueue::broadcast_fut_queue(64);
+		Consensus {
+			peers: HashMap::new(),
+			our_candidate: None,
+			statement_sink: statement_sink,
+			statement_stream: statement_stream,
+			bft_message_sink: bft_sink,
+			bft_message_stream: bft_stream,
+		}
+	}
+
+	/// Handle new connected peer.
+	pub fn new_peer(&mut self, _io: &mut SyncIo, _protocol: &Protocol, peer_id: PeerId, roles: &[message::Role]) {
+		if roles.iter().any(|r| *r == message::Role::Validator) {
+			trace!(target:"sync", "Registering validator {}", peer_id);
+			self.peers.insert(peer_id, PeerConsensus {
+				candidate_fetch: None,
+				candidate_available: None,
+			});
+		}
+	}
+
+	pub fn on_statement(&mut self, peer_id: PeerId, statement: message::Statement) {
+		if let Some(ref mut peer) = self.peers.get_mut(&peer_id) {
+			// TODO: validate signature?
+			match &statement.statement {
+				&message::UnsignedStatement::Candidate(ref receipt) => peer.candidate_available = Some(Hash::from(receipt.blake2_256())),
+				&message::UnsignedStatement::Available(ref hash) => peer.candidate_available = Some(*hash),
+				&message::UnsignedStatement::Valid(_) | &message::UnsignedStatement::Invalid(_) => (),
+			}
+			if let Err(e) = self.statement_sink.try_send(statement) {
+				trace!(target:"sync", "Error broadcasting statement notification: {:?}", e);
+			}
+		} else {
+			trace!(target:"sync", "Ignored statement from unregistered peer {}", peer_id);
+		}
+	}
+
+	pub fn statements(&self) -> multiqueue::BroadcastFutReceiver<message::Statement>{
+		self.statement_stream.add_stream()
+	}
+
+	pub fn on_bft_message(&mut self, peer_id: PeerId, message: message::BftMessage) {
+		if self.peers.contains_key(&peer_id) {
+			// TODO: validate signature?
+			if let Err(e) = self.bft_message_sink.try_send(message) {
+				trace!(target:"sync", "Error broadcasting BFT message notification: {:?}", e);
```

Comment:
dropping messages when the buffer is full will almost certainly cause consensus to fail

Reply:
The network handler should not block. If the buffer is full this means that the system just lack performance to  handle the messages and consensus will fail anyway. Blocking here will just move congestion to the lower level.

Reply:
I disagree -- we could get 65 messages all at once and then one will be dropped. But if those are the last messages we were going to receive then it's fine to let the agreement logic process them slowly. It isn't fine to drop the last commit message, for example.

Reply:
Increased the buffer to 65536 for now. Unfortunately it is preallocated.

Reply:
worrying but workable for PoC-1. This is something that should fail on OOM and not before

Reply:
I'll address that in the following PR. Probably ny replacing multiqueue with a vector of mpsc channels.